File: sdk/npm/src/providers/tempmail-lol-v2.ts
import { InternalEmailInfo, Email, Channel } from '../types';
import { normalizeEmail } from '../normalize';
import { fetchWithTimeout } from '../retry';

const CHANNEL: Channel = 'tempmail-lol-v2';
const API_BASE = 'https://api.tempmail.lol';

const HEADERS: Record<string, string> = {
  'Accept': 'application/json',
  'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36',
};

export async function generateEmail(): Promise<InternalEmailInfo> {
  const res = await fetchWithTimeout(`${API_BASE}/generate`, { headers: HEADERS });
  if (!res.ok) throw new Error(`tempmail-lol-v2: generate failed: ${res.status}`);
  const data = await res.json() as { address?: string; token?: string };
  if (!data.address || !data.token) throw new Error('tempmail-lol-v2: missing address or token');
  return { channel: CHANNEL, email: data.address, token: data.token };
}

export async function getEmails(token: string, email: string): Promise<Email[]> {
  const res = await fetchWithTimeout(`${API_BASE}/auth/${encodeURIComponent(token)}`, { headers: HEADERS });
  if (!res.ok) throw new Error(`tempmail-lol-v2: get emails failed: ${res.status}`);
  const data = await res.json() as { email?: any[] };
  const list = Array.isArray(data.email) ? data.email : [];
  return list.map((raw: any) => normalizeEmail({
    id: raw.id || raw._id || '',
    from: raw.from || raw.sender || '',
    to: email,
    subject: raw.subject || '',
    text: raw.body || raw.text || '',
    html: raw.html || raw.body || '',
    date: raw.date || raw.receivedAt || '',
    isRead: false,
  }, email));
}

File: sdk/rust/src/providers/tempmail_lol_v2.rs
/*!
 * tempmail.lol V2 渠道实现
 * API: https://api.tempmail.lol
 */

use serde_json::Value;
use crate::types::{Channel, EmailInfo, Email};
use crate::normalize::normalize_email;
use crate::config::{http_client, block_on, get_current_ua};

const API_BASE: &str = "https://api.tempmail.lol";

pub fn generate_email() -> Result<EmailInfo, String> {
    block_on(async {
        let resp = http_client()
            .get(format!("{}/generate", API_BASE))
            .header("User-Agent", get_current_ua())
            .header("Accept", "application/json")
            .send().await.map_err(|e| format!("tempmail-lol-v2 request failed: {}", e))?;

        if !resp.status().is_success() {
            return Err(format!("tempmail-lol-v2 generate failed: {}", resp.status()));
        }

        let data: Value = resp.json().await.map_err(|e| format!("parse failed: {}", e))?;
        let address = data["address"].as_str().unwrap_or("");
        let token = data["token"].as_str().unwrap_or("");

        if address.is_empty() || token.is_empty() {
            return Err("tempmail-lol-v2: missing address or token".into());
        }

        Ok(EmailInfo {
            channel: Channel::TempmailLolV2,
            email: address.to_string(),
            token: Some(token.to_string()),
            expires_at: None,
            created_at: None,
        })
    })
}

pub fn get_emails(token: &str, email: &str) -> Result<Vec<Email>, String> {
    let token = token.to_string();
    let email = email.to_string();
    block_on(async {
        let resp = http_client()
            .get(format!("{}/auth/{}", API_BASE, urlencoding::encode(&token)))
            .header("User-Agent", get_current_ua())
            .header("Accept", "application/json")
            .send().await.map_err(|e| format!("tempmail-lol-v2 request failed: {}", e))?;

        if !resp.status().is_success() {
            return Err(format!("tempmail-lol-v2 get emails failed: {}", resp.status()));
        }

        let data: Value = resp.json().await.map_err(|e| format!("parse failed: {}", e))?;
        let list = data["email"].as_array().cloned().unwrap_or_default();

        let mut out = Vec::with_capacity(list.len());
        for raw in &list {
            let from = raw["from"].as_str()
                .or_else(|| raw["sender"].as_str())
                .unwrap_or("");
            let flat = serde_json::json!({
                "id": raw["id"].as_str().or_else(|| raw["_id"].as_str()).unwrap_or(""),
                "from": from,
                "to": &email,
                "subject": raw["subject"].as_str().unwrap_or(""),
                "text": raw["body"].as_str().or_else(|| raw["text"].as_str()).unwrap_or(""),
                "html": raw["html"].as_str().or_else(|| raw["body"].as_str()).unwrap_or(""),
                "date": raw["date"].as_str().or_else(|| raw["receivedAt"].as_str()).unwrap_or(""),
                "isRead": false,
            });
            out.push(normalize_email(&flat, &email));
        }
        Ok(out)
    })
}
